crates/ra_hir_ty/src/_match.rs

```diff
@@ -0,0 +1,1174 @@
+}
+
+impl PatIdOrWild {
+    fn as_pat(self, cx: &MatchCheckCtx) -> Pat {
+        match self {
+            PatIdOrWild::PatId(id) => cx.body.pats[id].clone(),
+            PatIdOrWild::Wild => Pat::Wild,
+        }
+    }
+
+    fn as_id(self) -> Option<PatId> {
+        match self {
+            PatIdOrWild::PatId(id) => Some(id),
+            PatIdOrWild::Wild => None,
+        }
+    }
+}
+
+impl From<PatId> for PatIdOrWild {
+    fn from(pat_id: PatId) -> Self {
+        Self::PatId(pat_id)
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct MatchCheckNotImplemented;
+
+/// The return type of `is_useful` is either an indication of usefulness
+/// of the match arm, or an error in the case the match statement
+/// is made up of types for which exhaustiveness checking is currently
+/// not completely implemented.
+///
+/// The `std::result::Result` type is used here rather than a custom enum
+/// to allow the use of `?`.
+pub type MatchCheckResult<T> = Result<T, MatchCheckNotImplemented>;
+
+#[derive(Debug)]
+/// A row in a Matrix.
+///
+/// This type is modeled from the struct of the same name in `rustc`.
+pub(crate) struct PatStack(PatStackInner);
+type PatStackInner = SmallVec<[PatIdOrWild; 2]>;
+
+impl PatStack {
+    pub(crate) fn from_pattern(pat_id: PatId) -> PatStack {
+        Self(smallvec!(pat_id.into()))
+    }
+
+    pub(crate) fn from_wild() -> PatStack {
+        Self(smallvec!(PatIdOrWild::Wild))
+    }
+
+    fn from_slice(slice: &[PatIdOrWild]) -> PatStack {
+        Self(SmallVec::from_slice(slice))
+    }
+
+    fn from_vec(v: PatStackInner) -> PatStack {
+        Self(v)
+    }
+
+    fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    fn head(&self) -> PatIdOrWild {
+        self.0[0]
+    }
+
+    fn get_head(&self) -> Option<PatIdOrWild> {
+        self.0.first().copied()
+    }
+
+    fn to_tail(&self) -> PatStack {
+        Self::from_slice(&self.0[1..])
+    }
+
+    fn replace_head_with(&self, pat_ids: &[PatId]) -> PatStack {
+        let mut patterns: PatStackInner = smallvec![];
+        for pat in pat_ids {
+            patterns.push((*pat).into());
+        }
+        for pat in &self.0[1..] {
+            patterns.push(*pat);
+        }
+        PatStack::from_vec(patterns)
+    }
+
+    /// Computes `D(self)`.
+    ///
+    /// See the module docs and the associated documentation in rustc for details.
+    fn specialize_wildcard(&self, cx: &MatchCheckCtx) -> Option<PatStack> {
+        if matches!(self.head().as_pat(cx), Pat::Wild) {
+            Some(self.to_tail())
+        } else {
+            None
+        }
+    }
+
+    /// Computes `S(constructor, self)`.
+    ///
+    /// See the module docs and the associated documentation in rustc for details.
+    fn specialize_constructor(
+        &self,
+        cx: &MatchCheckCtx,
+        constructor: &Constructor,
+    ) -> MatchCheckResult<Option<PatStack>> {
+        let result = match (self.head().as_pat(cx), constructor) {
```

Comment:
can't this use `pat_constructor` to make the comparison a bit simpler?

Reply:
Hmm, this seems like a good idea, but I've attempted to implement it and for various reasons it ends up becoming more complicated than I would have expected. 

This may be pointing to the need for some deeper re-orginazation which is needed, or perhaps I just need to think about this one a bit more.

Reply:
Looking at this again, I think the biggest blocker for it is that `Constructor::Tuple` doesn't contain the `PatId` list, only the arity. This means we'd still need to match on the pattern, which doesn't seem nice, or replace the arity in `Constructor::Tuple` with a `Vec<PatId>`, which might be reasonable but goes against some discussions we had elsewhere in this PR about simplifying `Constructor`. 

For now I'd prefer to leave this as-is, knowing we can certainly re-visit this as we add functionality to this match checker (as this method will certainly grow). Does that seem reasonable?

Reply:
Yes, let's leave it as is for now.

I think it might make sense to separate the 'comparing the constructors' and 'expanding the pattern' parts. Comparing the constructors can then be done on `Constructor` alone, and expanding the pattern just needs to match on the pattern.